cloudferrylib/os/network/neutron.py

```diff
@@ -1081,81 +1102,70 @@ def add_router_interfaces(self, src_router, dst_router,
+            network.
+         2. If keep_floating_ips option is set:
+         2.1. Modify IP address of a floating IP to be the same as on
+              destination. This is done from the DB level.
+         2.2. Else - do not modify floating IP address
+        """
+        LOG.info("Uploading floating IPs...")
+        existing_networks = self.get_networks()
+        for fip in src_floats:
+            # keystone auth fails if done with token for some reason
+            ksclient = self.identity_client._get_client_by_creds()
+            with ksresource.AddAdminUserToNonAdminTenant(
+                    ksclient,
+                    self.config.cloud.user,
+                    fip['tenant_name']):
+
+                ext_net_id = self.get_new_extnet_id(
+                    fip['floating_network_id'], networks, existing_networks)
+
+                if ext_net_id is None:
+                    LOG.info("No external net for floating IP, make sure all "
+                             "external networks migrated. Skipping floating "
+                             "IP '%s'", fip['floating_ip_address'])
+                    continue
 
-        for src_float in src_floats:
-            tname = src_float['tenant_name']
-            tenant_id = \
-                self.identity_client.get_tenant_id_by_name(tname)
-            ext_net_id = self.get_new_extnet_id(src_float['floating_network_id'],
-                                                src_nets, existing_nets)
-            for floating in existing_floatingips:
-                tenants_on_src_and_dst_dont_match = (
-                    floating['floating_ip_address'] == src_float['floating_ip_address'] and
-                    floating['floating_network_id'] == ext_net_id and
-                    floating['tenant_id'] != tenant_id)
-
-                if tenants_on_src_and_dst_dont_match:
-                    fl_id = floating['id']
-
-                    LOG.debug("Moving FIP '%s' to tenant '%s (%s)'",
-                              floating['floating_ip_address'],
-                              tname, tenant_id)
-                    self.neutron_client.delete_floatingip(fl_id)
-                    self.neutron_client.create_floatingip({
-                        'floatingip':
-                        {
-                            'floating_network_id': ext_net_id,
-                            'tenant_id': tenant_id
-                        }
-                    })
+                tenant = ksclient.tenants.find(name=fip['tenant_name'])
+                new_fip = {
+                    'floatingip': {
+                        'floating_network_id': ext_net_id,
+                        'tenant_id': tenant.id
+                    }
+                }
+                LOG.debug("Creating FIP on net '%s'", ext_net_id)
+                created_fip = self.neutron_client.create_floatingip(new_fip)
+
+            if self.config.migrate.keep_floatingip:
```

Comment:
I guess we do not need this check here, cause it should be already done on upper level:
https://github.com/MirantisWorkloadMobility/CloudFerry/blob/devel/cloudferrylib/os/network/neutron.py#L79